### PR TITLE
DEX-1366 Removed 2nd step JWT validation

### PR DESCRIPTION
**Pull request by ngadiyak**

### CODE REVIEW

**Review by vsuharnikov — CHANGES_REQUESTED** [review]

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/protocol/WsAddressSubscribe.scala [hunk 1]
```diff
@@ -36,7 +52,7 @@ final case class WsAddressSubscribe(key: Address, authType: String, jwt: String,
         val given = payload.networkByte.head.toByte
         Either.cond(given == networkByte, (), error.TokenNetworkUnexpected(networkByte, given))
       }
-      _ <- Either.cond(crypto.verify(payload.signature, ByteStr(payload.toSign), payload.publicKey), (), error.InvalidJwtPayloadSignature)
+      payload <- Right(maybeWithDebug(payload))
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/ws/protocol/WsAddressSubscribe.scala]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/sync/api/ws/WsAddressStreamTestSuite.scala [hunk 2]
```diff
@@ -745,6 +745,23 @@ class WsAddressStreamTestSuite extends WsSuiteBase with TableDrivenPropertyCheck
...
+    Using.resource(mkDexWsConnection(dex1)) { wsc =>
+      wsc.send(
+        WsAddressSubscribe(
+          acc,
+          WsAddressSubscribe.defaultAuthType,
+          mkJwt(mkJwtNotSignedPayload(acc))
+        )
+      )
+
+      eventually(wsc.balanceChanges should have size 1)
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/sync/api/ws/WsAddressStreamTestSuite.scala]: fixed

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/protocol/WsAddressSubscribe.scala [hunk 3]
```diff
@@ -72,7 +92,8 @@ object WsAddressSubscribe {
     clientId: String,
     firstTokenExpirationInSeconds: Long,
     activeTokenExpirationInSeconds: Long,
-    scope: List[String]
+    scope: List[String],
+    isDebug: Boolean = false
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/ws/protocol/WsAddressSubscribe.scala:96]: `JwtPayload` is what we receive from an external service and it doesn't contain `isDebug`. Ideally we want to receive such information, but now it doesn't exist. So we must change `validate` to:
```scala
def validate(jwtPublicKey: String, networkByte: Byte): Either[MatcherError, (JwtPayload, Boolean)] =
```
where `true` means debug.

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/protocol/WsAddressSubscribe.scala [hunk 4]
```diff
@@ -15,9 +15,25 @@ import pdi.jwt.{JwtAlgorithm, JwtJson, JwtOptions}
...
+        publicKey = payload.publicKey,
+        networkByte = payload.networkByte,
+        clientId = payload.clientId,
+        firstTokenExpirationInSeconds = payload.firstTokenExpirationInSeconds,
+        activeTokenExpirationInSeconds = payload.activeTokenExpirationInSeconds,
+        scope = payload.scope,
+        isDebug = true
+      )
+
   def validate(jwtPublicKey: String, networkByte: Byte): Either[MatcherError, JwtPayload] =
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/ws/protocol/WsAddressSubscribe.scala:37]:
```scala
/**
  * @return true means that this token is a debug token
  */
def validate(jwtPublicKey: String, networkByte: Byte): Either[MatcherError, (JwtPayload, Boolean)] =
```
See below why

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/protocol/WsAddressSubscribe.scala [hunk 5]
```diff
@@ -15,9 +15,25 @@ import pdi.jwt.{JwtAlgorithm, JwtJson, JwtOptions}
 import play.api.libs.functional.syntax._
 import play.api.libs.json._
 
-final case class WsAddressSubscribe(key: Address, authType: String, jwt: String, flags: Set[WsAddressFlag] = Set.empty) extends WsClientMessage {
+final case class WsAddressSubscribe(private val key: Address, authType: String, jwt: String, flags: Set[WsAddressFlag] = Set.empty)
+    extends WsClientMessage {
   override val tpe: String = WsAddressSubscribe.tpe
 
+  private def maybeWithDebug(payload: JwtPayload): JwtPayload =
```
Comment (vsuharnikov) [inline dex/src/main/scala/com/wavesplatform/dex/api/ws/protocol/WsAddressSubscribe.scala:22]: This should be `private def isDebug(payload: JwtPayload): Boolean`. See below

**Review by vsuharnikov — CHANGES_REQUESTED** [review]

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala [hunk 6]
```diff
@@ -170,24 +170,29 @@ object WsExternalClientHandlerActor {
...
 
                 case subscribe: WsAddressSubscribe =>
-                  val address = subscribe.key
                   val authType = subscribe.authType
 
                   subscribe.validate(settings.jwtPublicKey, AddressScheme.current.chainId) match {
                     case Left(e) =>
-                      context.log.debug(s"WsAddressSubscribe(k=$address, t=$authType) failed with ${e.message.text}")
+                      context.log.debug(
+                        s"WsAddressSubscribe(k=${PublicKey.fromBase58String(settings.jwtPublicKey).getOrElse(PublicKey.empty).toAddress}, t=$authType) failed with ${e.message.text}"
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala:178]: we can't, address is private in WsAddressSubscribe

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala [hunk 7]
```diff
@@ -170,24 +170,29 @@ object WsExternalClientHandlerActor {
...
+                      )
                       clientRef ! WsError.from(e, matcherTime)
                       Behaviors.same
                     case Right(jwtPayload) =>
+                      if (jwtPayload.isDebug)
+                        context.log.debug(s"${jwtPayload.publicKey.toAddress} is connecting with debug token")
+
+                      val address = jwtPayload.publicKey.toAddress
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala:186]: moved upper

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala [hunk 8]
```diff
@@ -170,24 +170,29 @@ object WsExternalClientHandlerActor {
                   Behaviors.same
 
                 case subscribe: WsAddressSubscribe =>
-                  val address = subscribe.key
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala:173]: we can't, address is private in WsAddressSubscribe

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala [hunk 9]
```diff
@@ -170,24 +170,29 @@ object WsExternalClientHandlerActor {
...
                     case Left(e) =>
-                      context.log.debug(s"WsAddressSubscribe(k=$address, t=$authType) failed with ${e.message.text}")
+                      context.log.debug(
+                        s"WsAddressSubscribe(k=${PublicKey.fromBase58String(settings.jwtPublicKey).getOrElse(PublicKey.empty).toAddress}, t=$authType) failed with ${e.message.text}"
+                      )
                       clientRef ! WsError.from(e, matcherTime)
                       Behaviors.same
                     case Right(jwtPayload) =>
+                      if (jwtPayload.isDebug)
+                        context.log.debug(s"${jwtPayload.publicKey.toAddress} is connecting with debug token")
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala:184]: fixed

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/protocol/WsAddressSubscribe.scala (hunk 3 above)
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/ws/protocol/WsAddressSubscribe.scala:96]: fixed

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/protocol/WsAddressSubscribe.scala (hunk 5 above)
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/ws/protocol/WsAddressSubscribe.scala:22]: fixed

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala:173]: ok

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala:178]: fixed

File: dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala [hunk 10]
```diff
@@ -170,24 +170,29 @@ object WsExternalClientHandlerActor {
...
                       val expiration = scheduleOnce(subscriptionLifetime, CancelAddressSubscription(address))
 
                       addressSubscriptions
-                        .find(_._1 == subscribe.key)
+                        .find(_._1 == address)
                         .fold {
                           addressRef ! AddressDirectoryActor.Command.ForwardMessage(
-                            subscribe.key,
-                            AddressActor.WsCommand.AddWsSubscription(clientRef, subscribe.flags)
+                            address,
```
[the remark this answers is not in the document]
Comment (ngadiyak) [reply dex/src/main/scala/com/wavesplatform/dex/api/ws/actors/WsExternalClientHandlerActor.scala:195]: ok

**Pull request merged**